File: src/ships/assignToShipRole.ts
import type { CrewMember } from "../data/types/crewMember.ts";
import type { Ship } from "../data/types/ship.ts";
import type { ShipCrewRole } from "../data/types/shipCrewRole.ts";
import type { AssignShipRoleResult } from "../data/types/assignShipRoleResult.ts";
import { getCrewSlotsForShip } from "./getCrewSlotsForShip.ts";

// Ship Crew Roles amendment (profitable-design-questions.md). Mirrors
// assignToCraft()'s shape -- operates on an already-hired CrewMember,
// never touches hiring/pool/wage machinery. currentRoster is the
// player's full owned-crew list (any subset already assigned to *this*
// ship is what actually matters for the capacity count below).
export function assignToShipRole(
  crewMember: CrewMember,
  ship: Ship,
  role: ShipCrewRole,
  currentRoster: readonly CrewMember[],
): AssignShipRoleResult {
  // "The Crafter role slot reuses the existing profession taxonomy
  // directly" -- only tier-6/7 crew (profession !== null) are eligible.
  // The other 4 roles are explicitly NOT gated by tier or profession.
  if (role === "Crafter" && crewMember.profession === null) {
    return { assigned: false, reason: "only a crew member with a profession can be assigned the Crafter role" };
  }

  const slots = getCrewSlotsForShip(ship);
  // combatEngineerOrScienceOfficer is a single combined pool shared by
  // both roles, not two independent caps -- an occupant of either role
  // counts against the same total.
  const isCombinedPoolRole = role === "Combat Engineer" || role === "Science Officer";
  const capacity = isCombinedPoolRole
    ? slots.combatEngineerOrScienceOfficer
    : role === "Pilot"
      ? slots.pilot
      : role === "Systems Engineer"
        ? slots.systemsEngineer
        : slots.crafter;

  if (capacity <= 0) {
    return { assigned: false, reason: `no ${role} slot exists on a ${ship.tier}-tier ship` };
  }

  const occupied = currentRoster.filter(
    (member) =>
      member.id !== crewMember.id &&
      member.assignedShipId === ship.id &&
      (isCombinedPoolRole
        ? member.shipRole === "Combat Engineer" || member.shipRole === "Science Officer"
        : member.shipRole === role),
  ).length;

  if (occupied >= capacity) {
    return { assigned: false, reason: `${role} slot is full (${occupied}/${capacity})` };
  }

  return {
    assigned: true,
    // Reassigning clears the previous assignment automatically -- both
    // fields live on this same CrewMember record, so overwriting them
    // is the entire "clear the previous one" rule (Must-Not-Do: a crew
    // member can't hold two ship-role slots at once).
    updatedCrewMember: { ...crewMember, shipRole: role, assignedShipId: ship.id },
  };
}

File: unity/ProfitableCore/Simulation/AssignToShipRole.cs
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/ships/assignToShipRole.ts. Mirrors AssignToCraft's shape --
// operates on an already-hired CrewMember, never touches hiring/pool/
// wage machinery. CurrentRoster is the player's full owned-crew list.
public static class ShipRoleAssigner
{
    public static AssignShipRoleResult AssignToShipRole(CrewMember crewMember, Ship ship, ShipCrewRole role, IReadOnlyList<CrewMember> currentRoster)
    {
        // Only tier-6/7 crew (profession != null) are eligible for
        // Crafter. The other 4 roles are explicitly NOT gated by tier or
        // profession.
        if (role == ShipCrewRole.Crafter && crewMember.Profession is null)
        {
            return new AssignShipRoleRejected { Reason = "only a crew member with a profession can be assigned the Crafter role" };
        }

        var slots = ShipCrewSlotResolver.GetCrewSlotsForShip(ship);
        // CombatEngineerOrScienceOfficer is a single combined pool shared
        // by both roles, not two independent caps.
        var isCombinedPoolRole = role is ShipCrewRole.CombatEngineer or ShipCrewRole.ScienceOfficer;
        var capacity = isCombinedPoolRole
            ? slots.CombatEngineerOrScienceOfficer
            : role switch
            {
                ShipCrewRole.Pilot => slots.Pilot,
                ShipCrewRole.SystemsEngineer => slots.SystemsEngineer,
                _ => slots.Crafter,
            };

        if (capacity <= 0)
        {
            return new AssignShipRoleRejected { Reason = $"no {role} slot exists on a {ship.Tier}-tier ship" };
        }

        var occupied = currentRoster.Count(member =>
            member.Id != crewMember.Id &&
            member.AssignedShipId == ship.Id &&
            (isCombinedPoolRole
                ? member.ShipRole is ShipCrewRole.CombatEngineer or ShipCrewRole.ScienceOfficer
                : member.ShipRole == role));

        if (occupied >= capacity)
        {
            return new AssignShipRoleRejected { Reason = $"{role} slot is full ({occupied}/{capacity})" };
        }

        // Reassigning clears the previous assignment automatically --
        // both fields live on this same CrewMember record, so overwriting
        // them is the entire "clear the previous one" rule.
        var updatedCrewMember = new CrewMember
        {
            Id = crewMember.Id, HiredByPlayerId = crewMember.HiredByPlayerId, Tier = crewMember.Tier,
            Profession = crewMember.Profession, Status = crewMember.Status, AssignedCraftId = crewMember.AssignedCraftId,
            HiredAt = crewMember.HiredAt, LastCheckedAt = crewMember.LastCheckedAt, WageAmount = crewMember.WageAmount,
            LastPaidAt = crewMember.LastPaidAt, UnavailableUntil = crewMember.UnavailableUntil,
            ShipRole = role, AssignedShipId = ship.Id,
        };

        return new AssignShipRoleSucceeded { UpdatedCrewMember = updatedCrewMember };
    }
}
